index.js

```diff
@@ -1499,6 +1506,105 @@
     };
   }
 
+  //# validate :: Type -> a -> Array (Either Object a)
+  //.
+  //. Takes a type, and any value. Returns an `Array` of
+  //. `Right a` if the value is a member of the type;
+  //. `Left Object` for each property that is invalid.
+  //. The first index in an array of `Left`s is always named `$$`,
+  //. which refers to the entire value.
+  function validate(t) {
+    return function(x) {
+      // 1. build list of prop, validate function and value object for all keys
+      // 2. when value = null, use MissingValue object
+      // 3. run all remainder validate functions for nested type validation
+      // 4. remove all Rights from list
+      // 5. concatenate $$ (in returnValue) with all Lefts
+
+      var returnValue = [];
+      var $$Result    = t.validate ([]) (x);
+
+      // 1 and 2
+      //  props :: Array (Either Object Object)
+      var props = t.keys.map (function(p) {
+        return x == null
+          ? Left ({
+              error: 'MissingValue',
+              type: t.name || t.type,
+              name: p,
+              value: x
+            })
+          : Right ({
+              name: p,
+              type: t.types[p],
+              value: x[p]
+            });
+      });
+
+      var validateRights = Z.compose (function(p) {
+        if (p.result.isLeft) {
+          if (p.name in x) {
+            return Left ({
+              error: 'WrongValue',
+              // TODO: figure out what propPath really is
+              type: p.result.value.propPath.length > 0
+                ? p.type.types[p.result.value.propPath[0]].name
+                : p.type.name,
+              name: p.name,
+              value: p.value
+            });
```

Comment:
@davidchambers this part I'm really not sure about. The idea is to get the name of the Type that is invalid for nested types. I haven't seen any `propPath` that didn't exactly contain `['$1']` but I'm sure that there are cases.